docs/website/blog/2023-06-01-writing-devfile-from-scratch.mdx

```diff
@@ -0,0 +1,359 @@
+  ```yaml
+        kind: build
+        isDefault: true
+  ```
+  The previous block defines that this command belongs to `build` group and is the default command. 
+  Each group can have only one default command. When you run `odo dev`, odo automatically executes the default command in `build` group first,
+ followed by the default command in `run` group.
+ 
+ 
+Ideally, this would be all we need, and you could use this Devfile with odo.
+
+## Fixing issues
+If you try to use Devfile as we have it, you will see an error.
+The first problem is that the NodeJS image doesn't have `yarn` installed. 
+
+### Install `yaml` into the container
+To add yarn, we will leverage Devfile feature called [`events`](https://devfile.io/docs/2.2.0/adding-event-bindings).
+Events allow us to define commands that should be executed on predefined events. 
+There are 3 events that you can use.
+- `preStart` - executed before the main container is started.
+- `postStart` - executed after the main container is started.
+- `preStop` - executed before the main container is stopped.
+
+In our case we will use `preStart` event and execute `npm install -g yarn`.
+
+
+```yaml
+schemaVersion: 2.2.0
+metadata:
+  name: my-backstage
+
+commands:
+# highlight-start
+  - id: install-yarn
+    exec:
+      commandLine: npm install -g yarn
+      workingDir: ${PROJECT_SOURCE}
+      component: nodejs
+# highlight-end
+  - id: yarn-install
+    exec:
+      commandLine: npx yarn install
+      component: nodejs
+      workingDir: ${PROJECT_SOURCE}
+      group:
+        kind: build
+        isDefault: true
+    
+  - id: run-start
+    exec:
+      commandLine: npx yarn run dev
+      component: nodejs
+      workingDir: ${PROJECT_SOURCE}
+      group:
+        kind: run
+        isDefault: true
+    
+components:
+  - name: nodejs
+    container:
+      image: registry.access.redhat.com/ubi9/nodejs-18:latest
+      sourceMapping: /projects
+      command: ['tail', '-f', '/dev/null']
+      memoryLimit: 2Gi
+      endpoints:
+        - name: frontend
+          targetPort: 3000
+          exposure: public
+        - name: backend
+          targetPort: 7007
+          exposure: public
+# highlight-start
+events:
+  postStart:
+    - install-yarn
+# highlight-end
+```
+
+Even after installing `yarn` you won't be able to use this Devfile with odo and Backstage source code.
+
+
+### No space left on device 
+You will get `NOSPC: no space left on device` error.
+
+This is due to the [#6836](https://github.com/redhat-developer/odo/issues/6836) issue in odo.
+At the time of writing this, odo creates a 2GB volume for the source code. For Backstage and it's `node_modules`
+this is not enough. Luckily, there is a simple workaround that we can do in Devfile.
+
+We can create extra volume just for `/projects/node_modules`. This will put `node_modules` into a separate volume form the source code.
```

Comment:
```suggestion
We can create extra volume just for `/projects/node_modules`. This will put `node_modules` into a separate volume for the source code.
```

Reply:
It's possible the user does not see this error if he is using a persistent volume, as many provisioners provide volumes with a minimal size > 2Gb. For example on IBM Cloud, you get a 10Gb volume even if you request 2Gb.